Public/Src/Tools/Execution.Analyzer/Analyzers.Core/XLGPlusPlus/XLGToDBAnalyzer.cs

```diff
@@ -389,24 +470,155 @@ public override void DominoInvocation(DominoInvocationEventData data)
+                        xldbSpecificPip = modulePip.ToModulePip(PathTable, xldbPip);
+                    }
+                    else if (pipType == PipType.SealDirectory)
+                    {
+                        var sealDirectoryPip = (Pips.Operations.SealDirectory)hydratedPip;
+                        xldbSpecificPip = sealDirectoryPip.ToSealDirectory(PathTable, xldbPip);
+                    }
+                    else if (pipType == PipType.CopyFile)
+                    {
+                        var copyFilePip = (Pips.Operations.CopyFile)hydratedPip;
+                        xldbSpecificPip = copyFilePip.ToCopyFile(PathTable, xldbPip);
+                    }
+                    else if (pipType == PipType.WriteFile)
+                    {
+                        var writeFilePip = (Pips.Operations.WriteFile)hydratedPip;
+                        xldbSpecificPip = writeFilePip.ToWriteFile(PathTable, xldbPip);
+                    }
+                    else if (pipType == PipType.Process)
+                    {
+                        var processPip = (Pips.Operations.Process)hydratedPip;
+                        xldbSpecificPip = processPip.ToProcessPip(PathTable, xldbPip);
+                    }
+
+                    var pipSemistableQuery = new PipQuerySemiStableHash()
+                    {
+                        SemiStableHash = hydratedPip.SemiStableHash,
+                    };
+
+                    var pipSemistableValue = new PipValueSemiStableHash()
+                    {
+                        PipId = hydratedPip.PipId.Value,
+                    };
+
+                    var pipIdQuery = new PipQueryPipId()
+                    {
+                        PipId = hydratedPip.PipId.Value,
+                        PipType = (Xldb.PipType)pipType
+                    };
+
+                    // If the SemiStableHash != 0, then we want to create the SemistableHash -> PipId indirection.
+                    // Else we do not want that since the key would no longer be unique
+                    if (hydratedPip.SemiStableHash != 0)
+                    {
+                        pipSemistableKeys.Add(pipSemistableQuery.ToByteArray());
+                        pipSemistableValues.Add(pipSemistableValue.ToByteArray());
+                    }
+
+                    pipPipIdKeys.Add(pipIdQuery.ToByteArray());
+                    pipValues.Add(xldbSpecificPip.ToByteArray());
+                }
+
+                // Write the rest of the batched pips to the db
+                WriteBatchToDb(pipSemistableKeys, pipSemistableValues, XldbDataStore.PipColumnFamilyName);
+                WriteBatchToDb(pipPipIdKeys, pipValues, XldbDataStore.PipColumnFamilyName);
+            });
         }
 
         /// <summary>
         /// Write a key/value pair to the db
         /// </summary>
-        public void WriteToDb(byte[] key, byte[] value)
+        public void WriteToDb(byte[] key, byte[] value, string columnFamilyName = null)

```

Comment:
Noticed you have some usages like:

for (...) {
  ...
  WriteToDb(..)
}

This means you will take the lock in the accessor once for every write. I'd take it only once and pass the IBuildXLKeyValueStore around instead. Keep in mind that the accessor's lock is there only to avoid incorrect disposal of the DB.